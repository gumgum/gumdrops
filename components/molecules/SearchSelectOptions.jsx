import React, { Component } from 'react';
import PropTypes from 'prop-types';
import charCodes from '../../constants/charCodes';
import modulo from '../utils/modulo';

const styleReset = {
    display: 'block',
    width: '100%',
    border: 0,
    borderRadius: 0,
    textAlign: 'left',
    lineHeight: 'inherit',
    fontFamily: 'inherit',
    fontWeight: 'inherit',
    fontSize: 'inherit'
};

const NEXT = [charCodes.ARROW_RIGHT, charCodes.ARROW_DOWN];
const PREV = [charCodes.ARROW_LEFT, charCodes.ARROW_UP];

class SearchSelectOptions extends Component {
    static propTypes = {
        options: PropTypes.arrayOf(
            PropTypes.shape({
                id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
                title: PropTypes.string.isRequired
            })
        ),
        handleOptionClick: PropTypes.func.isRequired,
        noResultsMessage: PropTypes.string.isRequired
    };

    static defaultProps = {
        noResultsMessage: 'No results found'
    };

    state = {
        focusedIndex: 0
    };

    _handleFocus = currentIndex => event => {
        const { options } = this.props;

        if (NEXT.includes(event.keyCode)) {
            event.preventDefault();
            const nextIndex = modulo(currentIndex + 1, options.length);
            this[`_option${nextIndex}`].focus();
            return this.setState({ focusedIndex: nextIndex });
        }

        if (PREV.includes(event.keyCode)) {
            event.preventDefault();
            const prevIndex = modulo(currentIndex - 1, options.length);
            this[`_option${prevIndex}`].focus();
            return this.setState({ focusedIndex: prevIndex });
        }
    };

    renderOptions() {
        const { options, handleOptionClick, noResultsMessage } = this.props;
        const { focusedIndex } = this.state;

        return options.length ? (
            options.map(({ id, title }, index) => (
                <li key={id}>
                    <button
                        aria-label={title}
                        tabIndex={focusedIndex === index ? 0 : -1}
                        ref={ref => (this[`_option${index}`] = ref)}
                        className="gds-search-select__menu-item"
                        style={styleReset}
                        name={`option${index}`}
                        onKeyDown={this._handleFocus(index)}
                        onClick={() => handleOptionClick({ id, title })}>
                        {index} {title}
                    </button>
                </li>
            ))
        ) : (
            <li className="gds-search-select__menu-item">{noResultsMessage}</li>
        );
    }

    render() {
        return (
            <div className={`gds-search-select__menu`}>
                <ul className="gds-search-select__menu-items ">{this.renderOptions()}</ul>
            </div>
        );
    }
}

export default SearchSelectOptions;
