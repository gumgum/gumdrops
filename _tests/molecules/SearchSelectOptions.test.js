/* globals mount shallow */
import React from 'react';
import SearchSelectOptions from '../../components/molecules/SearchSelectOptions';
import charCodes from '../../constants/charCodes';

const options = [
    { id: 1, title: 'Lorem ipsum dolor sit amet' },
    { id: 2, title: 'Consectetur adipiscing elit ' },
    { id: 3, title: 'Sed do eiusmod tempor incididunt ut' },
    { id: 4, title: 'Labore et dolore magna aliqua' },
    { id: 5, title: 'Ut enim ad minim veniam' },
    { id: 6, title: 'Quis nostrud exercitation ullamco' },
    { id: 7, title: 'Laboris nisi ut aliquip ex ea commodo' },
    { id: 8, title: 'Consequat Duis' },
    { id: 9, title: 'Aute irure dolor in reprehenderit ' },
    { id: 10, title: 'In voluptate velit esse cillum dolore' },
    { id: 11, title: 'Eu fugiat nulla pariatur' },
    { id: 12, title: 'Deserunt mollit anim id est laborum' }
];

const defaultProps = {
    handleOptionClick: () => {}
};

describe('Expect <SearchSelectOptions>', () => {
    it('to render', () => {
        const wrapper = shallow(<SearchSelectOptions options={options} {...defaultProps} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('to render no results', () => {
        const wrapper = shallow(
            <SearchSelectOptions options={[]} {...defaultProps} noResultsMessage="oh no!" />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('to call handleOptionClick', () => {
        const handleOptionClick = jest.fn();
        const wrapper = shallow(
            <SearchSelectOptions options={options} handleOptionClick={handleOptionClick} />
        );
        const optionBtn = wrapper.find('button').first();
        optionBtn.simulate('click');

        expect(handleOptionClick).toBeCalledWith(options[0]);
    });

    it('to handle focus with arrow keys', () => {
        const wrapper = mount(<SearchSelectOptions options={options} {...defaultProps} />);
        [
            ['[name="option0"]', charCodes.ARROW_DOWN, 1],
            ['[name="option1"]', charCodes.ARROW_UP, 0],
            ['[name="option0"]', charCodes.ARROW_LEFT, 11],
            ['[name="option11"]', charCodes.ARROW_RIGHT, 0]
        ].forEach(([name, code, expected]) => {
            wrapper.find(name).simulate('keydown', { keyCode: code, preventDefault: () => {} });
            expect(wrapper.state().focusedIndex).toBe(expected);
        });
    });
});
