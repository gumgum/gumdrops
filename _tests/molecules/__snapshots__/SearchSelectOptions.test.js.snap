// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <SearchSelectOptions> to render 1`] = `
<div
  className="gds-search-select__menu"
>
  <ul
    className="gds-search-select__menu-items "
  >
    <li
      key="1"
    >
      <button
        aria-label="Lorem ipsum dolor sit amet"
        className="gds-search-select__menu-item"
        name="option0"
        onClick={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "border": 0,
            "borderRadius": 0,
            "display": "block",
            "fontFamily": "inherit",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "lineHeight": "inherit",
            "textAlign": "left",
            "width": "100%",
          }
        }
        tabIndex={0}
      >
        0
         
        Lorem ipsum dolor sit amet
      </button>
    </li>
    <li
      key="2"
    >
      <button
        aria-label="Consectetur adipiscing elit "
        className="gds-search-select__menu-item"
        name="option1"
        onClick={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "border": 0,
            "borderRadius": 0,
            "display": "block",
            "fontFamily": "inherit",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "lineHeight": "inherit",
            "textAlign": "left",
            "width": "100%",
          }
        }
        tabIndex={-1}
      >
        1
         
        Consectetur adipiscing elit 
      </button>
    </li>
    <li
      key="3"
    >
      <button
        aria-label="Sed do eiusmod tempor incididunt ut"
        className="gds-search-select__menu-item"
        name="option2"
        onClick={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "border": 0,
            "borderRadius": 0,
            "display": "block",
            "fontFamily": "inherit",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "lineHeight": "inherit",
            "textAlign": "left",
            "width": "100%",
          }
        }
        tabIndex={-1}
      >
        2
         
        Sed do eiusmod tempor incididunt ut
      </button>
    </li>
    <li
      key="4"
    >
      <button
        aria-label="Labore et dolore magna aliqua"
        className="gds-search-select__menu-item"
        name="option3"
        onClick={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "border": 0,
            "borderRadius": 0,
            "display": "block",
            "fontFamily": "inherit",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "lineHeight": "inherit",
            "textAlign": "left",
            "width": "100%",
          }
        }
        tabIndex={-1}
      >
        3
         
        Labore et dolore magna aliqua
      </button>
    </li>
    <li
      key="5"
    >
      <button
        aria-label="Ut enim ad minim veniam"
        className="gds-search-select__menu-item"
        name="option4"
        onClick={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "border": 0,
            "borderRadius": 0,
            "display": "block",
            "fontFamily": "inherit",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "lineHeight": "inherit",
            "textAlign": "left",
            "width": "100%",
          }
        }
        tabIndex={-1}
      >
        4
         
        Ut enim ad minim veniam
      </button>
    </li>
    <li
      key="6"
    >
      <button
        aria-label="Quis nostrud exercitation ullamco"
        className="gds-search-select__menu-item"
        name="option5"
        onClick={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "border": 0,
            "borderRadius": 0,
            "display": "block",
            "fontFamily": "inherit",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "lineHeight": "inherit",
            "textAlign": "left",
            "width": "100%",
          }
        }
        tabIndex={-1}
      >
        5
         
        Quis nostrud exercitation ullamco
      </button>
    </li>
    <li
      key="7"
    >
      <button
        aria-label="Laboris nisi ut aliquip ex ea commodo"
        className="gds-search-select__menu-item"
        name="option6"
        onClick={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "border": 0,
            "borderRadius": 0,
            "display": "block",
            "fontFamily": "inherit",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "lineHeight": "inherit",
            "textAlign": "left",
            "width": "100%",
          }
        }
        tabIndex={-1}
      >
        6
         
        Laboris nisi ut aliquip ex ea commodo
      </button>
    </li>
    <li
      key="8"
    >
      <button
        aria-label="Consequat Duis"
        className="gds-search-select__menu-item"
        name="option7"
        onClick={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "border": 0,
            "borderRadius": 0,
            "display": "block",
            "fontFamily": "inherit",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "lineHeight": "inherit",
            "textAlign": "left",
            "width": "100%",
          }
        }
        tabIndex={-1}
      >
        7
         
        Consequat Duis
      </button>
    </li>
    <li
      key="9"
    >
      <button
        aria-label="Aute irure dolor in reprehenderit "
        className="gds-search-select__menu-item"
        name="option8"
        onClick={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "border": 0,
            "borderRadius": 0,
            "display": "block",
            "fontFamily": "inherit",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "lineHeight": "inherit",
            "textAlign": "left",
            "width": "100%",
          }
        }
        tabIndex={-1}
      >
        8
         
        Aute irure dolor in reprehenderit 
      </button>
    </li>
    <li
      key="10"
    >
      <button
        aria-label="In voluptate velit esse cillum dolore"
        className="gds-search-select__menu-item"
        name="option9"
        onClick={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "border": 0,
            "borderRadius": 0,
            "display": "block",
            "fontFamily": "inherit",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "lineHeight": "inherit",
            "textAlign": "left",
            "width": "100%",
          }
        }
        tabIndex={-1}
      >
        9
         
        In voluptate velit esse cillum dolore
      </button>
    </li>
    <li
      key="11"
    >
      <button
        aria-label="Eu fugiat nulla pariatur"
        className="gds-search-select__menu-item"
        name="option10"
        onClick={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "border": 0,
            "borderRadius": 0,
            "display": "block",
            "fontFamily": "inherit",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "lineHeight": "inherit",
            "textAlign": "left",
            "width": "100%",
          }
        }
        tabIndex={-1}
      >
        10
         
        Eu fugiat nulla pariatur
      </button>
    </li>
    <li
      key="12"
    >
      <button
        aria-label="Deserunt mollit anim id est laborum"
        className="gds-search-select__menu-item"
        name="option11"
        onClick={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "border": 0,
            "borderRadius": 0,
            "display": "block",
            "fontFamily": "inherit",
            "fontSize": "inherit",
            "fontWeight": "inherit",
            "lineHeight": "inherit",
            "textAlign": "left",
            "width": "100%",
          }
        }
        tabIndex={-1}
      >
        11
         
        Deserunt mollit anim id est laborum
      </button>
    </li>
  </ul>
</div>
`;

exports[`Expect <SearchSelectOptions> to render no results 1`] = `
<div
  className="gds-search-select__menu"
>
  <ul
    className="gds-search-select__menu-items "
  >
    <li
      className="gds-search-select__menu-item"
    >
      oh no!
    </li>
  </ul>
</div>
`;
